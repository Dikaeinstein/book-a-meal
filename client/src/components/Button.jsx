import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ value, className }) => (
  <button className={className}>
    {value}
  </button>
);

Button.propTypes = {
  value: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
};

export default Button;
