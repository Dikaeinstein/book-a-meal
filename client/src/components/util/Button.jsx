import React from 'react';
import PropTypes from 'prop-types';

const Button = ({ value, ...otherProps }) => (
  <button {...otherProps}>
    {value}
  </button>
);

Button.propTypes = {
  value: PropTypes.string.isRequired,
  className: PropTypes.string.isRequired,
};

export default Button;
