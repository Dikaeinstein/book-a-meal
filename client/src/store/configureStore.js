import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { createLogger } from 'redux-logger';
import throttle from 'lodash.throttle';
import { loadState, saveState } from '../helpers/localStorage';
import initialState from '../reducers/initialState';
import rootReducer from '../reducers';

/**
 * Return redux middleware
 */
const getMiddleware = () => {
  const logger = createLogger();

  const middleware =
    process.env.NODE_ENV !== 'production'
    && process.env.NODE_ENV !== 'test'
      ? applyMiddleware(thunk, logger)
      : applyMiddleware(thunk);

  return middleware;
};

/**
 * Configure the redux store
 *
 * @returns {object} Redux store
 */
const configureStore = () => {
  const persistedState = loadState();

  const state = persistedState
    ? { ...initialState, ...persistedState }
    : initialState;

  const store = createStore(
    rootReducer,
    state,
    compose(
      getMiddleware(),
      window.devToolsExtension ? window.devToolsExtension() : empty => empty,
    ),
  );

  // Throttle the saveState function to be called once every second
  store.subscribe(throttle(() => {
    saveState({ orders: store.getState().orders });
  }, 1000));

  return store;
};

export default configureStore;
