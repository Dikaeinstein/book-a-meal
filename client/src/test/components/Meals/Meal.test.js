import React from 'react';
import { shallow } from 'enzyme';
import sweetalert from 'sweetalert';
import configureStore from 'redux-mock-store';
import ConnectedMeal, { Meal } from '../../../components/Meals/Meal';
import Button from '../../../components/util/Button';

jest.mock('sweetalert');

describe('<Meal />', () => {
  const meal = {
    name: 'Test meal',
    description: 'This is a test meal',
    user: { name: 'Test Dika' },
    imageUrl: 'http://testimageurl',
    price: 4000,
  };
  const props = {
    meal,
    isDeleting: false,
    isUpdating: false,
    handleMealUpdate: jest.fn(),
    removeMeal: jest.fn(),
    removeCatererMeal: jest.fn(),
    userRole: 'superAdmin',
  };
  it('should call handleOpenModal when viewMealDetail is clicked', () => {
    const openModalSpy = jest.spyOn(Meal.prototype, 'handleOpenModal');
    const wrapper = shallow(<Meal {...props} />);
    const viewMealDetail = wrapper.find('.overlay-container');
    viewMealDetail.simulate('click');
    expect(openModalSpy).toHaveBeenCalled();
    expect(wrapper.state().isOpen).toEqual(true);
  });
  it('should call handleCloseModal when viewMealDetail is clicked', () => {
    const closeModalSpy = jest.spyOn(Meal.prototype, 'handleCloseModal');
    const wrapper = shallow(<Meal {...props} />);
    const closeMealDetail = wrapper.find(Button);
    closeMealDetail.simulate('click');
    expect(closeModalSpy).toHaveBeenCalled();
    expect(wrapper.state().isOpen).toEqual(false);
  });
  it('should call handleDelete when "Delete" button is clicked', () => {
    const handleDeleteSpy = jest.spyOn(Meal.prototype, 'handleDelete');
    let wrapper = shallow(<Meal {...props} />);
    let deleteMeal = wrapper.find('button').last();

    sweetalert.mockResolvedValue(Promise.resolve(true));
    deleteMeal.simulate('click');
    expect(handleDeleteSpy).toHaveBeenCalled();

    sweetalert.mockResolvedValue(Promise.reject(new Error()));
    deleteMeal.simulate('click');
    expect(wrapper).toMatchSnapshot();

    sweetalert.mockResolvedValue(Promise.reject());
    deleteMeal.simulate('click');
    expect(wrapper).toMatchSnapshot();

    wrapper = shallow(<Meal {...{ ...props, userRole: 'caterer' }} />);
    deleteMeal = wrapper.find('button').last();
    sweetalert.mockResolvedValue(Promise.resolve(true));
    deleteMeal.simulate('click');
    expect(handleDeleteSpy).toHaveBeenCalled();
  });
  it('should call handlUpdate when "Edit" button is clicked', () => {
    const handleUpdateSpy = jest.spyOn(Meal.prototype, 'handleUpdate');

    let wrapper = shallow(<Meal {...props} />);
    let editMeal = wrapper.find('button').first();
    sweetalert.mockResolvedValue(Promise.resolve(true));
    editMeal.simulate('click');
    expect(handleUpdateSpy).toHaveBeenCalled();

    wrapper = shallow(<Meal {...{ ...props, userRole: 'caterer' }} />);
    editMeal = wrapper.find('button').first();

    sweetalert.mockResolvedValue(Promise.resolve(true));
    editMeal.simulate('click');
    expect(wrapper).toMatchSnapshot();
  });
  it('should render with correct props from redux', () => {
    const mockStore = configureStore();
    const store = mockStore({
      meals: { isUpdating: false, isDeleting: false },
      user: { data: { role: 'superAdmin' } },
    });
    const wrapper = shallow(<ConnectedMeal {...props} store={store} />);
    expect(wrapper.props().isUpdating).toEqual(false);
    expect(wrapper.props().isDeleting).toEqual(false);
  });
  it('should match snapshot', () => {
    const tree = shallow(<Meal {...props} />);
    expect(tree).toMatchSnapshot();
  });
});
