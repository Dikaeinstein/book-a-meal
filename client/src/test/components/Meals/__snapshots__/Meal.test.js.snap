// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Meal /> should call handlUpdate when "Edit" button is clicked 1`] = `
<div
  className="card user-menu-item"
>
  <div
    className="overlay-container"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <img
      alt="Test meal"
      src="http://testimageurl"
    />
    <div
      className="overlay"
    >
      <div
        className="text"
      >
        <p>
          View description
        </p>
      </div>
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Meal Detail"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "70%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <MealDetail
      closeMealDetail={[Function]}
      meal={
        Object {
          "description": "This is a test meal",
          "imageUrl": "http://testimageurl",
          "name": "Test meal",
          "price": 4000,
          "user": Object {
            "name": "Test Dika",
          },
        }
      }
    />
  </Modal>
  <div
    className="font-weight-bold text-black text-center"
    style={
      Object {
        "padding": ".875rem",
      }
    }
  >
    <p>
      Test meal
    </p>
    <p
      style={
        Object {
          "fontSize": "1.25rem",
        }
      }
    >
      ₦
      4000
    </p>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "margin": ".5rem auto",
        "width": "75%",
      }
    }
  >
    <button
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#28a745",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Edit Meal"
      value="Edit"
    >
      Edit
    </button>
    <button
      disabled={false}
      onClick={[Function]}
      resource="meal"
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#dc3545",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Delete Meal"
      value="Delete"
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`<Meal /> should call handleDelete when "Delete" button is clicked 1`] = `
<div
  className="card user-menu-item"
>
  <div
    className="overlay-container"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <img
      alt="Test meal"
      src="http://testimageurl"
    />
    <div
      className="overlay"
    >
      <div
        className="text"
      >
        <p>
          View description
        </p>
      </div>
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Meal Detail"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "70%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <MealDetail
      closeMealDetail={[Function]}
      meal={
        Object {
          "description": "This is a test meal",
          "imageUrl": "http://testimageurl",
          "name": "Test meal",
          "price": 4000,
          "user": Object {
            "name": "Test Dika",
          },
        }
      }
    />
  </Modal>
  <div
    className="font-weight-bold text-black text-center"
    style={
      Object {
        "padding": ".875rem",
      }
    }
  >
    <p>
      Test meal
    </p>
    <p
      style={
        Object {
          "fontSize": "1.25rem",
        }
      }
    >
      ₦
      4000
    </p>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "margin": ".5rem auto",
        "width": "75%",
      }
    }
  >
    <button
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#28a745",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Edit Meal"
      value="Edit"
    >
      Edit
    </button>
    <button
      disabled={false}
      onClick={[Function]}
      resource="meal"
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#dc3545",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Delete Meal"
      value="Delete"
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`<Meal /> should call handleDelete when "Delete" button is clicked 2`] = `
<div
  className="card user-menu-item"
>
  <div
    className="overlay-container"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <img
      alt="Test meal"
      src="http://testimageurl"
    />
    <div
      className="overlay"
    >
      <div
        className="text"
      >
        <p>
          View description
        </p>
      </div>
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Meal Detail"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "70%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <MealDetail
      closeMealDetail={[Function]}
      meal={
        Object {
          "description": "This is a test meal",
          "imageUrl": "http://testimageurl",
          "name": "Test meal",
          "price": 4000,
          "user": Object {
            "name": "Test Dika",
          },
        }
      }
    />
  </Modal>
  <div
    className="font-weight-bold text-black text-center"
    style={
      Object {
        "padding": ".875rem",
      }
    }
  >
    <p>
      Test meal
    </p>
    <p
      style={
        Object {
          "fontSize": "1.25rem",
        }
      }
    >
      ₦
      4000
    </p>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "margin": ".5rem auto",
        "width": "75%",
      }
    }
  >
    <button
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#28a745",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Edit Meal"
      value="Edit"
    >
      Edit
    </button>
    <button
      disabled={false}
      onClick={[Function]}
      resource="meal"
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#dc3545",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Delete Meal"
      value="Delete"
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`<Meal /> should match snapshot 1`] = `
<div
  className="card user-menu-item"
>
  <div
    className="overlay-container"
    onClick={[Function]}
    onKeyPress={[Function]}
    role="button"
    tabIndex="0"
  >
    <img
      alt="Test meal"
      src="http://testimageurl"
    />
    <div
      className="overlay"
    >
      <div
        className="text"
      >
        <p>
          View description
        </p>
      </div>
    </div>
  </div>
  <Modal
    ariaHideApp={true}
    bodyOpenClassName="ReactModal__Body--open"
    closeTimeoutMS={0}
    contentLabel="Meal Detail"
    isOpen={false}
    parentSelector={[Function]}
    portalClassName="ReactModalPortal"
    shouldCloseOnEsc={true}
    shouldCloseOnOverlayClick={true}
    shouldFocusAfterRender={true}
    shouldReturnFocusAfterClose={true}
    style={
      Object {
        "content": Object {
          "bottom": "auto",
          "left": "50%",
          "marginRight": "-50%",
          "minWidth": "20rem",
          "right": "auto",
          "top": "50%",
          "transform": "translate(-50%, -50%)",
          "width": "70%",
        },
        "overlay": Object {
          "backgroundColor": "rgba(0, 0, 0, 0.25)",
        },
      }
    }
  >
    <Button
      className="close"
      onClick={[Function]}
      value="×"
    />
    <MealDetail
      closeMealDetail={[Function]}
      meal={
        Object {
          "description": "This is a test meal",
          "imageUrl": "http://testimageurl",
          "name": "Test meal",
          "price": 4000,
          "user": Object {
            "name": "Test Dika",
          },
        }
      }
    />
  </Modal>
  <div
    className="font-weight-bold text-black text-center"
    style={
      Object {
        "padding": ".875rem",
      }
    }
  >
    <p>
      Test meal
    </p>
    <p
      style={
        Object {
          "fontSize": "1.25rem",
        }
      }
    >
      ₦
      4000
    </p>
  </div>
  <div
    style={
      Object {
        "display": "flex",
        "margin": ".5rem auto",
        "width": "75%",
      }
    }
  >
    <button
      disabled={false}
      onClick={[Function]}
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#28a745",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Edit Meal"
      value="Edit"
    >
      Edit
    </button>
    <button
      disabled={false}
      onClick={[Function]}
      resource="meal"
      style={
        Object {
          "background": "#e9ebeb",
          "border": "none",
          "borderRadius": "3px",
          "color": "#dc3545",
          "cursor": "pointer",
          "flex": "50%",
          "margin": ".5rem",
          "padding": ".5rem",
        }
      }
      title="Delete Meal"
      value="Delete"
    >
      Delete
    </button>
  </div>
</div>
`;
